keepalived-vip/utils.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"errors"
+	"net"
+	"regexp"
+	"sort"
+)
+
+var (
+	invalidIfaces = []string{"lo", "docker0", "flannel.1", "cbr0"}
+	vethRegex     = regexp.MustCompile(`^veth.*`)
+)
+
+type networkInfo struct {
+	iface   string
+	ip      string
+	netmask int
+}
+
+// getNetworkInfo returns information of the node where the pod is running
+func getNetworkInfo(ip string) (*networkInfo, error) {
+	iface, _, mask := interfaceByIP(ip)
+	return &networkInfo{
+		iface:   iface,
+		ip:      ip,
+		netmask: mask,
+	}, nil
+}
+
+// netInterfaces returns a slice containing the local network interfaces
+// excluding lo, docker0, flannel.1 and veth interfaces.
+func netInterfaces() []net.Interface {
+	validIfaces := []net.Interface{}
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return validIfaces
+	}
+
+	for _, iface := range ifaces {
+		if !vethRegex.MatchString(iface.Name) && stringSlice(invalidIfaces).pos(iface.Name) == -1 {
+			validIfaces = append(validIfaces, iface)
+		}
+	}
+
+	return validIfaces
+}
+
+// interfaceByIP returns the local network interface name that is using the
+// specified IP address. If no interface is found returns an empty string.
+func interfaceByIP(ip string) (string, string, int) {
+	for _, iface := range netInterfaces() {
+		ifaceIP, mask, err := ipByInterface(iface.Name)
+		if err == nil && ip == ifaceIP {
+			return iface.Name, ip, mask
+		}
+	}
+
+	return "", "", 0
+}
+
+func ipByInterface(name string) (string, int, error) {
+	iface, err := net.InterfaceByName(name)
+	if err != nil {
+		return "", 32, err
+	}
+
+	addrs, err := iface.Addrs()
+	if err != nil {
+		return "", 32, err
+	}
+
+	for _, a := range addrs {
+		if ipnet, ok := a.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+			if ipnet.IP.To4() != nil {
+				ip := ipnet.IP.String()
+				ones, _ := ipnet.Mask.Size()
+				mask := ones
+				return ip, mask, nil
+			}
+		}
+	}
+
+	return "", 32, errors.New("Found no IPv4 addresses.")
+}
+
+// getPriority returns the priority of one node using the
+// IP address as key. It starts in 100
+func getPriority(ip string, peers []string) int {
```

Comment:
I sort peers, and priority=100+ (index of peer in the array). No particular rationale, just because keepalived requires peers must have their priorities.